fri/src/verifier.rs

```diff
@@ -129,35 +128,70 @@ where
     let mut folded_eval = F::ZERO;
     let mut ro_iter = reduced_openings.into_iter().peekable();
 
-    for (log_folded_height, (&beta, comm, opening)) in izip!((0..log_max_height).rev(), steps) {
-        if let Some((_, ro)) = ro_iter.next_if(|(lh, _)| *lh == log_folded_height + 1) {
+    let mut log_folded_height = log_max_height;
+
+    while log_folded_height > config.log_blowup + config.log_final_poly_len {
+        if let Some((_, ro)) = ro_iter.next_if(|(lh, _)| *lh == log_folded_height) {
             folded_eval += ro;
         }
 
-        let index_sibling = index ^ 1;
-        let index_pair = index >> 1;
+        let (&beta, comm, opening) = steps.next().unwrap();
+
+        let index_row = index >> config.arity_bits;
+
+        // Verify that `folded_eval` and evals from reduced_openings match opened_rows
+        assert_eq!(folded_eval, opening.opened_rows[0][index % config.arity()]);
+        for row in opening.opened_rows.iter().skip(1) {
+            // TODO[osama]: consider adding an assert for the length of row to make sure the right thing is being read
+            let (lh, ro) = ro_iter.next().unwrap();
+            let current_index = index >> (log_folded_height - lh);
+            assert_eq!(
+                ro,
+                row[current_index % row.len()],
+                "reduced opening mismatch at level {:?}",
+                lh
+            );
+        }
 
-        let mut evals = vec![folded_eval; 2];
-        evals[index_sibling % 2] = opening.sibling_value;
+        let dims: Vec<_> = opening
+            .opened_rows
+            .iter()
+            .map(|opened_row| Dimensions {
+                width: opened_row.len(),
+                height: 1 << (log_folded_height - config.arity_bits),
+            })
+            .collect();
 
-        let dims = &[Dimensions {
-            width: 2,
-            height: 1 << log_folded_height,
-        }];
         config
             .mmcs
             .verify_batch(
                 comm,
-                dims,
-                index_pair,
-                &[evals.clone()],
+                &dims,
+                index_row,
+                &opening.opened_rows,
                 &opening.opening_proof,
             )
             .map_err(FriError::CommitPhaseMmcsError)?;
 
-        index = index_pair;
+        // Do the folding logic
+
+        let mut opened_rows_iter = opening.opened_rows.iter().peekable();
+        let mut folded_row = opened_rows_iter.next().unwrap().clone();
+        let mut index_folded_row = index_row << config.arity_bits;
+        while folded_row.len() > 1 {
+            index >>= 1;
+            log_folded_height -= 1;
```

Comment:
I'm not totally sure if it applies here, but again beware that two matrices could have had the same height